datajunction-server/datajunction_server/models/node.py

```diff
@@ -234,8 +360,96 @@ class AvailabilityStateBase(BaseSQLModel):
     schema_: Optional[str] = Field(default=None)
     table: str
     valid_through_ts: int
-    max_partition: List[str] = Field(sa_column=SqlaColumn(JSON))
-    min_partition: List[str] = Field(sa_column=SqlaColumn(JSON))
+
+    # An ordered list of categorical partitions like ["country", "group_id"]
+    # or ["region_id", "age_group"]
+    categorical_partitions: Optional[List[str]] = Field(
+        sa_column=SqlaColumn("categorical_partitions", JSON),
+        default=[],
+    )
+
+    # An ordered list of temporal partitions like ["date", "hour"] or ["date"]
+    temporal_partitions: Optional[List[str]] = Field(
+        sa_column=SqlaColumn("temporal_partitions", JSON),
+        default=[],
+    )
+
+    # Node-level temporal ranges
+    min_temporal_partition: Optional[List[str]] = Field(
+        sa_column=SqlaColumn(JSON),
+        default=[],
+    )
+    max_temporal_partition: Optional[List[str]] = Field(
+        sa_column=SqlaColumn(JSON),
+        default=[],
+    )
+
+    # Partition-level availabilities
+    partitions: Optional[List[PartitionAvailability]] = Field(
+        sa_column=SqlaColumn("partitions", JSON),
+        default=[],
+    )
+
+    @validator("partitions")
+    def validate_partitions(cls, partitions):  # pylint: disable=no-self-argument
+        """
+        Validator for partitions
+        """
+        return [partition.dict() for partition in partitions] if partitions else []
+
+    def merge(self, other: "AvailabilityStateBase"):
+        """
+        Merge this availability state with another.
+        """
+        all_partitions = [
+            PartitionAvailability(**partition)
+            if isinstance(partition, dict)
+            else partition
+            for partition in self.partitions + other.partitions  # type: ignore
+        ]
+        top_level_partition = PartitionAvailability(
+            value=[None for _ in other.categorical_partitions]
+            if other.categorical_partitions
+            else [],
+            min_temporal_partition=min(
+                x
+                for x in (self.min_temporal_partition, other.min_temporal_partition)
+                if x
+            ),
+            max_temporal_partition=max(
+                x
+                for x in (self.max_temporal_partition, other.max_temporal_partition)
+                if x
+            ),
+        )
+        all_partitions += [top_level_partition]
+
+        tracker = AvailabilityTracker()
+        for partition in all_partitions:
+            tracker.insert(partition)
+        final_partitions = tracker.get_partition_range()
+
+        self.partitions = [
+            partition
+            for partition in final_partitions
+            if not all(val is None for val in partition.value)
+        ]
+        merged_top_level = [
```

Comment:
Will this work as well if we only have temporal partitions on a nod and no categorical ones?